### PR TITLE
policy: remove `CIDRPolicy` structure

### LINKED ISSUE #6372: policy: look into removing ResolveCIDRPolicy functionality
This function is currently used to generate the list of allowed CIDRs per-endpoint. It does an iteration over all rules in the repository, which is costly in terms of matching on EndpointSelectors. 

It is also used to determine the number of prefix lengths of the CIDRs which apply. @joestringer suggested we could potentially just calculate a global list of prefix-lengths across all rules, instead of per-endpoint. This may increase complexity of programs for endpoints on older kernels, since not all the rules may select all endpoints, but it would increase the scalability of policy calculation itself. This only affects kernels which do not have LPM (4.9-4.11 IIRC).

@joestringer let me know if I synthesized our discussion about this correctly here, and feel free to add any additional context.

### PR DESCRIPTION
**Pull request by ianvernon**
While resolving policy with a large amount of rules containing CIDRs, a significant amount of time was taken running `resolveCIDRPolicy`:
    
```
(pprof) list resolvePolicyLocked
Total: 3.58s
ROUTINE ======================== github.com/cilium/cilium/pkg/policy.(*Repository).resolvePolicyLocked in /Users/ianvernon/go/src/github.com/cilium/cilium/pkg/policy/repository.go
             0      2.72s (flat, cum) 75.98% of Total
             .          .    590:   // First obtain whether policy applies in both traffic directions, as well
             .          .    591:   // as list of rules which actually select this endpoint. This allows us
             .          .    592:   // to not have to iterate through the entire rule list multiple times and
             .          .    593:   // perform the matching decision again when computing policy for each
             .          .    594:   // protocol layer, which is quite costly in terms of performance.
             .       70ms    595:   ingressEnabled, egressEnabled, matchingRules := p.computePolicyEnforcementAndRules(securityIdentity)
             .          .    596:
             .          .    597:   calculatedPolicy := &selectorPolicy{
             .          .    598:           Revision:             p.GetRevision(),
             .          .    599:           SelectorCache:        p.GetSelectorCache(),
             .          .    600:           L4Policy:             NewL4Policy(p.GetRevision()),
             .       10ms    601:           CIDRPolicy:           NewCIDRPolicy(),
             .          .    602:           IngressPolicyEnabled: ingressEnabled,
             .          .    603:           EgressPolicyEnabled:  egressEnabled,
             .          .    604:   }
             .          .    605:   calculatedPolicy.IngressPolicyEnabled = ingressEnabled
             .          .    606:   calculatedPolicy.EgressPolicyEnabled = egressEnabled
             .          .    607:
             .          .    608:   labels := securityIdentity.LabelArray
             .          .    609:   ingressCtx := SearchContext{
             .          .    610:           To:          labels,
             .          .    611:           rulesSelect: true,
             .          .    612:   }
             .          .    613:
             .          .    614:   egressCtx := SearchContext{
             .          .    615:           From:        labels,
             .          .    616:           rulesSelect: true,
             .          .    617:   }
             .          .    618:
             .          .    619:   if option.Config.TracingEnabled() {
             .          .    620:           ingressCtx.Trace = TRACE_ENABLED
             .          .    621:           egressCtx.Trace = TRACE_ENABLED
             .          .    622:   }
             .          .    623:
             .          .    624:   if ingressEnabled {
             .          .    625:           newL4IngressPolicy, err := matchingRules.resolveL4IngressPolicy(&ingressCtx, p.GetRevision(), p.GetSelectorCache())
             .          .    626:           if err != nil {
             .          .    627:                   return nil, err
             .          .    628:           }
             .          .    629:
             .          .    630:           newCIDRIngressPolicy := matchingRules.resolveCIDRPolicy(&ingressCtx)
             .          .    631:           if err := newCIDRIngressPolicy.Validate(); err != nil {
             .          .    632:                   return nil, err
             .          .    633:           }
             .          .    634:
             .          .    635:           calculatedPolicy.CIDRPolicy.Ingress = newCIDRIngressPolicy.Ingress
             .          .    636:           calculatedPolicy.L4Policy.Ingress = newL4IngressPolicy
             .          .    637:   }
             .          .    638:
             .          .    639:   if egressEnabled {
             .      2.02s    640:           newL4EgressPolicy, err := matchingRules.resolveL4EgressPolicy(&egressCtx, p.GetRevision(), p.GetSelectorCache())
             .          .    641:           if err != nil {
             .          .    642:                   return nil, err
             .          .    643:           }
             .          .    644:
             .      620ms    645:           newCIDREgressPolicy := matchingRules.resolveCIDRPolicy(&egressCtx)
             .          .    646:           if err := newCIDREgressPolicy.Validate(); err != nil {
             .          .    647:                   return nil, err
             .          .    648:           }
             .          .    649:
             .          .    650:           calculatedPolicy.CIDRPolicy.Egress = newCIDREgressPolicy.Egress
```
Approximately 17% of policy calculation time is taken by `resolveCIDRPolicy` in this case. Now that we compute all CIDR-based policy in terms of selectors and identities allocated for CIDRs, this function does not have any purpose from a policy perspective. Its main purposes were for adding visbility into CIDR-based policy in the Endpoint API, and for obtaining the set of prefix lengths for kernels >= 4.9 <= 4.11. These prefix lengths are needed for these kernel versions because said versions do not support LPM map types, and the implementation of the ipcache in the datapath varies in complexity based off of the number of prefix lengths of CIDRs used for policy. The prefix lengths of the resultant CIDR-based policy for each endpoint to determine if the number of prefix lengths returned were greater than the empirically measured number of prefix lengths which were determined to break the upper-bound of the complexity limit in the BPF verifier for these kernels.

Instead, we now use the number of unique prefix lengths in the ipcache itself as the number of prefixes for IPv6 and IPv4 for which the endpoint's BPF program is aware, which is global to the entire daemon, instead of per-CIDR policy in the endpoint. If policy is imported into cilium which does not select an endpoint, the endpoint will not be aware of these new prefix lengths until it is next regenerated. The CIDR-based representation in the endpoint API is now empty, but is still accessible in Kubernetes environments via CEP.
    
Total policy generation benchmark time before this change:
    
```
PASS: resolve_test.go:214: PolicyTestSuite.BenchmarkRegenerateCIDRPolicyRules        500           4541126 ns/op
```
    
Total policy generation benchmark time after this change:
    
```
PASS: resolve_test.go:214: PolicyTestSuite.BenchmarkRegenerateCIDRPolicyRules        500           3403568 ns/op
```
    
The performance improvement is approximately 25 % with this change, which will help a lot in clusters with policies containing large numbers of CIDRs.

Signed-off by: Ian Vernon <ian@cilium.io>

Considerations to make:

* This may confuse some end-users about the state of how CIDRPolicy is realized within the API. Our representation of policy in the API is somewhat incomplete right now anyway ( see #8506 ). 
* ~~The number of unique prefix lengths is now global vs. endpoint-local, which could (?) lead to failed builds for endpoints running   4.9 <= kernel version <= 4.11 which did not fail before - I'd like to clearly document whether there are any expected changes in behavior here, please comment @joestringer . Overall, I think that we need to reconsider whether we should bump the minimal kernel version we support in Cilium since Debian is now running a newer kernel version, most end-users we have interacted with are using newer kernels, etc.~~ - addressed below. 

Fixes: #6372 

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8469)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

File: pkg/datapath/fake/datapath.go [hunk 1]
```diff
@@ -17,22 +17,29 @@ package fake
...
-		node:           NewNodeHandler(),
-		nodeAddressing: NewNodeAddressing(),
+		node:                NewNodeHandler(),
+		nodeAddressing:      NewNodeAddressing(),
+		prefixLengthCounter: newPrefixLengthCounter(),
 	}
 }
 
+func newPrefixLengthCounter() *counter.PrefixLengthCounter {
+	return counter.NewPrefixLengthCounter(4, 18)
```
Comment (ianvernon) [inline pkg/datapath/fake/datapath.go:40]: @joestringer I used the constants from `pkg/maps/ipcache/ipcache.go` here but didn't want to import BPF-related code here... let me know if you want me to do more refactoring of the constants, happy to do so.

Comment (joestringer) [reply pkg/datapath/fake/datapath.go:40]: We could always refactor the constants into `pkg/defaults`.

File: test/runtime/Policies.go [hunk 2]
```diff
@@ -259,6 +259,11 @@ var _ = Describe("RuntimePolicies", func() {
 
 	BeforeAll(func() {
 		vm = helpers.InitRuntimeHelper(helpers.Runtime, logger)
+
+		// Make sure that Cilium is started with appropriate CLI options
+		// (specifically to exclude the local addresses that are populated for
+		// CIDR policy tests).
+		Expect(vm.SetUpCilium()).To(BeNil())
```
Comment (jrajahalme) [inline test/runtime/Policies.go:266]: This and the following change seem independent from this PR, and may need to be backported to older releases? As such maybe split these to a separate PR?

**Review by jrajahalme — APPROVED** [review]
Small nit only.

**Comment by joestringer** [comment]
There should be no change to CIDR prefix length handling. If we import more than the maximum, it should still be rejected at the policy layer before we attempt to use all of those prefixes for a given endpoint regeneration.

**Comment by joestringer** [comment]
> This may confuse some end-users about the state of how CIDRPolicy is realized within the API. Our representation of policy in the API is somewhat incomplete right now anyway ( see #8506 ).

We stripped out all references from the user docs that refer to CIDRPolicy anyway, so I don't think there's much left to confuse.

File: test/runtime/Policies.go (hunk 2 above)
Comment (ianvernon) [reply test/runtime/Policies.go:266]: Created https://github.com/cilium/cilium/pull/8548

**Comment by ianvernon** [comment]
test-me-please
